iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -25,111 +29,230 @@ namespace Flow {
+    // Block count mismatch. We do this here so that we avoid the O(n) scan
+    // that would have been required to calculate the size above.
+    return false;
+  }
+  return true;
 }
 
-bool areRegionsEquivalent(Region *lhs, Region *rhs) {
-  if (lhs->getBlocks().size() != rhs->getBlocks().size()) {
-    return false;
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs,
+                                       BlockAndValueMapping &parentMapping);
+static bool isStructurallyEquivalentTo(Operation &lhs, Operation &rhs,
+                                       BlockAndValueMapping &parentMapping);
+
+// Recursively compares two regions for structural equivalence.
+// Structural equivalence ensures that operations on both the |lhs| and |rhs|
+// have the same attributes and same use-def structure.
+//
+// Example:
+//   func @lhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//   func @rhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//
+//   assert(isStructurallyEquivalentTo(lhs.getBody(), rhs.getBody()));
+//
+// TODO(#3996): upstream into mlir::OperationEquivalence if this works.
+// TODO(#3996): add symbol ref comparison (add to BlockAndValueMapping).
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs) {
+  BlockAndValueMapping mapping;
+  return isStructurallyEquivalentTo(lhs, rhs, mapping);
+}
+
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs,
+                                       BlockAndValueMapping &mapping) {
+  // Use compare_ranges to walk the block list in parallel and get a boolean in
+  // the case of size mismatch without an O(N) linked-list size query.
+  if (!compare_ranges(
+          lhs.getBlocks(), rhs.getBlocks(),
+          [&](Block &lhsBlock, Block &rhsBlock) {
+            if (lhsBlock.getNumArguments() != rhsBlock.getNumArguments()) {
+              return false;
+            }
+            for (auto argPair :
+                 llvm::zip(lhsBlock.getArguments(), rhsBlock.getArguments())) {
+              auto &lhsArg = std::get<0>(argPair);
+              auto &rhsArg = std::get<1>(argPair);
+              if (lhsArg.getType() != rhsArg.getType()) return false;
+              mapping.map(lhsArg, rhsArg);
+            }
+            mapping.map(&lhsBlock, &rhsBlock);
+            return true;
+          })) {
+    return false;  // block mismatch
   }
 
-  for (auto blockPair : llvm::zip(lhs->getBlocks(), rhs->getBlocks())) {
+  // Walk the blocks again now that we have a populated mapping.
+  // We do this in topological order so that we have all values required by a
+  // block mapped by the time we reach it observing transitive block dominance.
+  llvm::SetVector<Block *> lhsBlocks;
+  for (Block &b : lhs.getBlocks()) {
+    if (lhsBlocks.count(&b) == 0) {
```

Comment:
is there a reason you are doing this outer for loop + if? It feels like just an RPO traversal of the entry is sufficient. (if not, it's worth a comment).

Reply:
No reason on my part - the code I used as a reference does this, so I did it too :)
https://github.com/llvm/llvm-project/blob/db884dafb7b5771e6ae01e8252f1520fac3e1c77/mlir%2Flib%2FTarget%2FLLVMIR%2FModuleTranslation.cpp#L381-L386